### ISSUE TITLE
GitCommitBear: Require the commit message body contains a URL pattern

### ISSUE THREAD

**Issue by jayvdb** [coala/coala-bears#1112]
[Writing Good Commits](https://api.coala.io/en/latest/Developers/Writing_Good_Commits.html) requires a URL.

We should have a bear that checks this.

**Comment by nkprince007** [comment]
Assign this to me

**Comment by jayvdb** [comment]
Could you join the gitter channel, so we can add you to the org, and assign you to this issue

**Comment by jayvdb** [comment]
As noted at https://github.com/coala/coala-bears/pull/1115/files#r92322299 (which could disappear)..

Enabling this feature is going to cause gitmate to reject any "multiple commits PR" where each commit does not have an issue linked. i.e. it is going to reject quite a few of our commits. e.g. coala/coala@ea48d66 coala/coala@49f8884 coala/coala@49f441b . coala/coala@39b0a17 .
Are we ready for this?

**Comment by jayvdb** [comment]
Interesting; [BitBucket commit issue linker](https://confluence.atlassian.com/bitbucket/resolve-issues-automatically-when-users-push-code-221451126.html) doesnt appear to support links, and my limited testing of it confirms this.  I've created https://github.com/coala/coala-bears/issues/1134 about that.

**Comment by jayvdb** [comment]
I like the fact the current implementation uses the actual github/gitlab rules, and that should be the default.

However, the origin site should be automatically detected, using `.git/config`.

Then someone can activate a more specific pattern if they want to.

**Comment by jayvdb** [comment]
https://github.com/coala/coala-bears/pull/1115 has all the code we need, with lots of features, but the way they are enabled in the settings needs a bit of design.

When we look at https://github.com/coala/bear-docs/blob/master/docs/GitCommitBear.rst , we see most of the settings are either `body_` or `shortlog_`.

There is already `shortlog_regex`, which can be used to require `Fixes #123`, and of course it is pretty silly to add URLs to the shortlog.

With the regex mode that is currently in https://github.com/coala/coala-bears/pull/1115 , it is essentially adding the complimenting `body_regex` .  To be a proper `body_regex: str` , it really needs to be a multiline regex.

The auto-closing checker functionality (i.e. https://help.github.com/articles/closing-issues-via-commit-messages/) could be enabled with a setting `body_close_issue: bool`.  (This is sort of https://github.com/coala/coala-bears/issues/1134 ).  This feature would only be supported on hosts that can be detected by `.git/config`, and is not strict about where it appears.

We could also have `body_close_issue_last_line: bool`, to force the issue close voodoo to appear on the last line only.

Then the requirement to use a URL (a coala specific requirement) would be is `body_close_issue_url: bool`.

**Comment by nkprince007** [comment]
coala/coala#3255 contains issue related to the aforementioned naming conflicts. And we definitely do need to have a naming scheme section in [Writing Bears](https://api.coala.io/en/latest/Developers/Writing_Bears.html) ASAP.

**Comment by Makman2** [comment]
> The auto-closing checker functionality (i.e. https://help.github.com/articles/closing-issues-via-commit-messages/) could be enabled with a setting body_close_issue: bool. (This is sort of #1134 ). This feature would only be supported on hosts that can be detected by .git/config, and is not strict about where it appears.

I believe you mean configured remotes inside `.git/config`, I think we should not look at those remotes, that is imo not really important^^ Just extract links from the commit and see whether it contains a valid issue reference.

**Comment by jayvdb** [comment]
A valid issue reference URL depends on the hostname of the issue tracker, which is why locating the remote helps auto-detection of the auto-closing functionality.  Also GitHub and GitLab support different sets of keywords for auto-closing.

**Comment by Makman2** [comment]
omg yes obviously, I've totally missed that you can shortcut issues
```
Closes #1234
or
Closes https://github.com/coala/coala/issues/1234
```
